File: DesignPatterns/Behavioral/Observer/NumberSystemClient.java
import java.util.ArrayList;
import java.util.List;

abstract class Subject {
  protected List<Observer> obsList;

  public Subject() {
    obsList = new ArrayList<Observer>();
  }

  protected void registerObserver(Observer obs){
    obsList.add(obs);
  }

  protected void unregisterObserver(Observer obs) {
    obsList.remove(obsList.indexOf(obs));
  }
  public abstract void notifyObservers();
}

class NumberSystemSubject extends Subject {
  private int value;

  public NumberSystemSubject(int value){
    super();
    this.value = value;
  }

  public void notifyObservers(){
    for(Observer obs: obsList) {
      obs.update(this.value);
    }
  }
}

interface Observer{
  public void update(int value);
}

class HexObserver implements Observer {
  public void update(int value){
    System.out.println(" " + Integer.toHexString(value));
  }
}

class OctObserver implements Observer {
  public void update(int value) {
    System.out.println(" " + Integer.toOctalString(value));
  }
}
class BinObserver implements Observer {
  public void update(int value) {
    System.out.println(" " + Integer.toBinaryString(value));
  }
}

public class NumberSystemClient {
  public static void main(String[] args) {
    HexObserver hex = new HexObserver();
    OctObserver oct = new OctObserver();
    BinObserver bin = new BinObserver();

    Subject sub = new NumberSystemSubject(55);
    sub.registerObserver(hex);
    sub.registerObserver(oct);
    sub.notifyObservers();

    System.out.println("************");
    sub.registerObserver(bin);
    sub.notifyObservers();
  }
}

File: DesignPatterns/Behavioral/Observer/Ruby/number_system_client.rb
class Subject
  attr_reader :observers

  def initialize
    @observers = []
  end

  def register_observer(observer)
    @observers << observer
  end

  def unregister_observer(observer)
    @observers.delete_if { |obs| obs == observer }
  end
end

class NumberSystemSubject < Subject
  attr_reader :value

  def initialize(value)
    super()
    @value = value
  end

  def notify_observers
    @observers.each do |observer|
      observer.update(@value)
    end
  end
end

class Observer
  def update(value)
    raise NotImplementedError
  end
end

class HexObserver < Observer
  def update(value)
    puts " #{value.to_s(16)}"
  end
end

class OctObserver < Observer
  def update(value)
    puts " #{value.to_s(8)}"
  end
end

class BinObserver < Observer
  def update(value)
    puts " #{value.to_s(2)}"
  end
end

hex = HexObserver.new
oct = OctObserver.new
bin = BinObserver.new

subject = NumberSystemSubject.new(55)
subject.register_observer(hex)
subject.register_observer(oct)

subject.notify_observers

puts "*************"
subject.unregister_observer(oct)
subject.register_observer(bin)
subject.notify_observers